File: 763_partition_labels.py
'''
A string S of lowercase letters is given. We want to partition this string into as many parts as possible so that each letter appears in at most one part, and return a list of integers representing the size of these parts.

Example 1:
Input: S = "ababcbacadefegdehijhklij"
Output: [9,7,8]
Explanation:
The partition is "ababcbaca", "defegde", "hijhklij".
This is a partition so that each letter appears in at most one part.
A partition like "ababcbacadefegde", "hijhklij" is incorrect, because it splits S into less parts.
Note:

S will have length in range [1, 500].
S will consist of lowercase letters ('a' to 'z') only.
'''

class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        last_idx = {}
        for i, char in enumerate(S):
            last_idx[char] = i
        
        s = 0
        e = 0
        parts = []
        for i, char in enumerate(S):
            if last_idx[char] > e:
                e = last_idx[char]
            if i == e:
                parts.append(e-s+1)
                s = e+1

        return parts

File: 763_partition_labels.go
/*
You are given a string s. We want to partition the string into as many parts as possible so that each letter appears in at most one part.

Note that the partition is done so that after concatenating all the parts in order, the resultant string should be s.

Return a list of integers representing the size of these parts.

 

Example 1:

Input: s = "ababcbacadefegdehijhklij"
Output: [9,7,8]
Explanation:
The partition is "ababcbaca", "defegde", "hijhklij".
This is a partition so that each letter appears in at most one part.
A partition like "ababcbacadefegde", "hijhklij" is incorrect, because it splits s into less parts.
Example 2:

Input: s = "eccbbbbdec"
Output: [10]
 

Constraints:

1 <= s.length <= 500
s consists of lowercase English letters.
*/

func partitionLabels(s string) []int {
    start := 0
    end := 0
    res := []int{}
    lastIndex := make([]int, 26)
    for i, c := range s {
        lastIndex[c - 'a'] = i
    }
    for i, c := range s {
        end = max(end, lastIndex[c - 'a'])
        if i == end {
            res = append(res, end - start + 1)
            start = end + 1
        }
    }
    return res
}

func max(a, b int) int {
    if a > b { return a }
    return b
}

